Makefile

```diff
@@ -1,32 +1,67 @@
-
 export CGO_ENABLED:=0
-LD_FLAGS="-w -X github.com/coreos/matchbox/matchbox/version.Version=$(shell ./git-version)"
-LOCAL_BIN=/usr/local/bin
 
-all: build
-build: clean bin/matchbox bin/bootcmd
+VERSION=$(shell ./scripts/git-version)
+LD_FLAGS="-w -X github.com/coreos/matchbox/matchbox/version.Version=$(VERSION)"
 
-tools:
-	./scripts/gentools
+REPO=github.com/coreos/matchbox
+IMAGE_REPO=coreos/matchbox
+QUAY_REPO=quay.io/coreos/matchbox
 
-codegen: tools
-	./scripts/codegen
+all: build
+
+build: clean bin/matchbox bin/bootcmd
 
 bin/matchbox:
-	go build -o bin/matchbox -ldflags $(LD_FLAGS) -a github.com/coreos/matchbox/cmd/matchbox
+	@go build -o bin/matchbox -v -ldflags $(LD_FLAGS) $(REPO)/cmd/matchbox
 
 bin/bootcmd:
-	go build -o bin/bootcmd -ldflags $(LD_FLAGS) -a github.com/coreos/matchbox/cmd/bootcmd
+	@go build -o bin/bootcmd -v -ldflags $(LD_FLAGS) $(REPO)/cmd/bootcmd
 
 test:
-	./test
+	@./scripts/test
+
+.PHONY: aci
+aci: clean build
+	@sudo ./scripts/build-aci
+
+.PHONY: docker-image
+docker-image:
+	@sudo docker build --rm=true -t $(IMAGE_REPO):$(VERSION) .
+	@sudo docker tag $(IMAGE_REPO):$(VERSION) $(IMAGE_REPO):latest
+
+.PHONY: docker-push
+docker-push: docker-image
+	@sudo docker tag $(IMAGE_REPO):$(VERSION) $(QUAY_REPO):latest
+	@sudo docker tag $(IMAGE_REPO):$(VERSION) $(QUAY_REPO):$(VERSION)
+	@sudo docker push $(QUAY_REPO):latest
+	@sudo docker push $(QUAY_REPO):$(VERSION)
+
+.PHONY: vendor
+vendor:
+	@glide update --strip-vendor
+	@glide-vc --use-lock-file --no-tests --only-code
+
+.PHONY: codegen
+codegen: tools
+	@./scripts/codegen
 
-install:
-	cp bin/matchbox $(LOCAL_BIN)
-	cp bin/bootcmd $(LOCAL_BIN)
+.PHONY: tools
+tools: bin/protoc bin/protoc-gen-go
+
+bin/protoc:
+	@./scripts/get-protoc
+
+bin/protoc-gen-go:
+	@go build -o bin/protoc-gen-go $(REPO)/vendor/github.com/golang/protobuf/protoc-gen-go
+
+clean:
+	@rm -rf bin
```

Comment:
I prefer to explicitly remove each file from the build directory and then remove the directory only if it is empty. This only works if make knows about every generated file, so this might not work here.